vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java

```diff
@@ -351,8 +363,28 @@ private void reset() {
         keyMapper.removeAll();
         removeAll();
         clear();
-        getDataProvider().fetch(new Query<>()).map(this::createCheckBox)
-                .forEach(this::add);
+
+        synchronized (dataProvider) {
+            final AtomicInteger itemCounter = new AtomicInteger(0);
+            getDataProvider().fetch(new Query<>()).map(this::createCheckBox)
+                    .forEach(component -> {
+                        add(component);
+                        itemCounter.incrementAndGet();
+                    });
+            lastFetchedDataSize = itemCounter.get();
+
+            // Ignore new size requests unless the last one has been executed
+            // so as to avoid multiple beforeClientResponses.
+            if (sizeRequest == null) {
+                sizeRequest = context -> {
```

Comment:
`ui` not `context`
My mistake as I was thinking of using `SerializableConsumer<ExecutionContext>` originally, but you had `UI` in the method.

Reply:
Also changing to ExecutionContext we could have `.beforeClientResponse(stateNode, command)`